lldb/examples/formatter-bytecode/compiler.py

```diff
@@ -0,0 +1,486 @@
+"""
+Specification, compiler, disassembler, and interpreter
+for LLDB dataformatter bytecode.
+
+See formatter-bytecode.md for more details.
+"""
+from __future__ import annotations
+
+# Types
+type_String = 1
+type_Int = 2
+type_UInt = 3
+type_Object = 4
+type_Type = 5
+
+# Opcodes
+opcode = dict()
+
+
+def define_opcode(n, mnemonic, name):
+    globals()["op_" + name] = n
+    if mnemonic:
+        opcode[mnemonic] = n
+    opcode[n] = mnemonic
+
+
+define_opcode(1, "dup", "dup")
+define_opcode(2, "drop", "drop")
+define_opcode(3, "pick", "pick")
+define_opcode(4, "over", "over")
+define_opcode(5, "swap", "swap")
+define_opcode(6, "rot", "rot")
+
+define_opcode(0x10, "{", "begin")
+define_opcode(0x11, "if", "if")
+define_opcode(0x12, "ifelse", "ifelse")
+
+define_opcode(0x20, None, "lit_uint")
+define_opcode(0x21, None, "lit_int")
+define_opcode(0x22, None, "lit_string")
+define_opcode(0x23, None, "lit_selector")
+
+define_opcode(0x30, "+", "plus")
+define_opcode(0x31, "-", "minus")
+define_opcode(0x32, "*", "mul")
+define_opcode(0x33, "/", "div")
+define_opcode(0x34, "%", "mod")
+define_opcode(0x35, "<<", "shl")
+define_opcode(0x36, ">>", "shr")
+define_opcode(0x37, "shra", "shra")
+
+define_opcode(0x40, "&", "and")
+define_opcode(0x41, "|", "or")
+define_opcode(0x42, "^", "xor")
+define_opcode(0x43, "~", "not")
+
+define_opcode(0x50, "=", "eq")
+define_opcode(0x51, "!=", "neq")
+define_opcode(0x52, "<", "lt")
+define_opcode(0x53, ">", "gt")
+define_opcode(0x54, "=<", "le")
+define_opcode(0x55, ">=", "ge")
+
+define_opcode(0x60, "call", "call")
+
+# Function signatures
+sig_summary = 0
+sig_init = 1
+sig_get_num_children = 2
+sig_get_child_index = 3
+sig_get_child_at_index = 4
+
+# Selectors
+selector = dict()
```

Comment:
I'm not familiar with the term "selector" in this context. Is it inspired by an existing thing that I can go read about?

It feels like "member function", like I would call "get_type" a "member function" of a value object.

Reply:
But then you have `sprintf` which isn't a member of anything it's a globally defined thing.

Reply:
But the point of this fixed list is to prevent a security issue if you could call anything.

Do these selectors run as if their `this` pointer is the ValueObject that you're currently formatting? So it's a bit like you don't have a global scope, that global scope is instead the methods of the ValueObject, plus a few utilities. So your "context", your world view, is just that one ValueObject.

Reply:
I lifted the word _selector_ from Objective-C (https://en.wikipedia.org/wiki/Objective-C#Messages) to describe the concept of an identifier for a method.

Reply:
It's an (intentionally!) hardcoded list of functions the program can call.
```
argn ... arg1 arg0 @selector call
```
behaves like the equivalent C code
```
selector(arg0, arg1, ... argn)
```
And in the cases of the ValueObj methods arg0 is the implicit `this` argument.